Capture logs of crashed pod

Jaanki opened the pull request:
Capture logs of crashed pod

by using Previous (-p) option of PodLogOptions

The logic implemented is

```
Capture the logs with -p
if previous_pod_found
    write_logs_to_file.log.prev
else
   ignore

Capture the logs without -p on the same pod
write_logs_to_file.log
```


Closes: #1289

Signed-Off-By: Janki Chhatbar <jchhatba@redhat.com>

mangelajo commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -57,7 +57,7 @@ func gatherPodLogs(podLabelSelector string, info Info) {
 
 func podLogsToFile(pod *corev1.Pod, info Info) error {
-	logRequest := info.ClientSet.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &corev1.PodLogOptions{})
+	logRequest := info.ClientSet.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &corev1.PodLogOptions{Previous:true})
```
We should only use Previous:true if the request with previous:false fails.

mangelajo replied:
I guess that we need to try L62 with previous=false, ... if that fails, then we try again with Previous:true

vthapar replied:
Why not get both? We could have scenarios where a pod crashed, caused some failures but then recovered. In such cases we will want old and current logs both.

Jaanki replied:
This patch does that.

I have a operator pod that restarted once.
`
submariner-operator       submariner-operator-7979fd59bb-psg98             1/1     Running     1          35h`

And this is reflected in the output
```
⚠ Gathering operator logs 
 ✓ Found 1 pods matching label selector "name=submariner-operator"
 ⚠ Found logs for previous instances of pod submariner-operator-7979fd59bb-psg98
```
and created
cluster2_submariner-operator-7979fd59bb-psg98.log
cluster2_submariner-operator-7979fd59bb-psg98.log.prev

mangelajo commented:
@Jaanki thanks for looking at this so quickly!

Jaanki commented:
GW pod in my setup is in `CrashLoopBackOff`. This patch generates a `.log.old` file with logs from previous pod and `.log` file with logs from current pod.

Attached are the logs files generated on running `subctl gather --module=connectivity --type=logs` 

[submariner-20210423170957.zip](https://github.com/submariner-io/submariner-operator/files/6367207/submariner-20210423170957.zip)

Jaanki committed:
Capture logs of crashed pod

by using Previous (-p) option of PodLogOptions

The logic implemented is

Capture the logs with -p
if previous_pod_found
    write_logs_to_file.log.prev
else
   ignore

Capture the logs without -p on the same pod
write_logs_to_file.log

Closes: #1289

Signed-Off-By: Janki Chhatbar <jchhatba@redhat.com>

Jaanki commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -62,41 +61,62 @@ func gatherPodLogsByContainer(podLabelSelector, container string, info Info) {
-	logs, err := processLogStream(logRequest)
+	// if no previous pods found, logstream == nil, ignore it
+	if logStream != nil {
+		info.Status.QueueWarningMessage(fmt.Sprintf("found previous instances of pod %s", pod.Name))
```
This warning message can be omitted

mangelajo replied:
It's ok to keep I think.

mangelajo commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -62,41 +61,62 @@ func gatherPodLogsByContainer(podLabelSelector, container string, info Info) {
-func podLogsToFile(pod *corev1.Pod, podLogOptions *corev1.PodLogOptions, info Info) error {
+func getlogs(pod *corev1.Pod, podLogOptions *corev1.PodLogOptions, info Info) error {
+	podLogOptions.Previous = true
 	logRequest := info.ClientSet.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, podLogOptions)
+	logStream, _ := logRequest.Stream()  // TODO: Check for error other than "no previous pods found"
 
-	logs, err := processLogStream(logRequest)
+	// if no previous pods found, logstream == nil, ignore it
+	if logStream != nil {
+		info.Status.QueueWarningMessage(fmt.Sprintf("found previous instances of pod %s", pod.Name))
+		err := writeLogs(logStream, info, pod.Name, ".log.old")
+		if err != nil {
+			return err
+		}
+		defer logStream.Close()
+	}
```
let's make this part a function getPreviousLogs ...

mangelajo commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -62,41 +61,62 @@ func gatherPodLogsByContainer(podLabelSelector, container string, info Info) {
+		defer logStream.Close()
+	}
+
+	// Running with Previous = false on the same pod
+	var err error
+	podLogOptions.Previous = false
+	logRequest = info.ClientSet.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, podLogOptions)
+	logStream, err = logRequest.Stream()
+	if err != nil {
+		return errors.WithMessage(err, "error opening log stream")
+	}
+	defer logStream.Close()
+	err = writeLogs(logStream, info, pod.Name, ".log")
+	if err != nil {}
+	return nil
```
And this part another function:  "getExistingLogs()"

In such way we divide the function in smaller more meaningful chunks

mangelajo reviewed (changes requested):
Just a nit, looks good

tpantelis commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -62,41 +61,46 @@ func gatherPodLogsByContainer(podLabelSelector, container string, info Info) {
 
-func podLogsToFile(pod *corev1.Pod, podLogOptions *corev1.PodLogOptions, info Info) error {
-	logRequest := info.ClientSet.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, podLogOptions)
+func getlogs(pod *corev1.Pod, podLogOptions *corev1.PodLogOptions, info Info) error {
```
This should be camel-case but why was this renamed? `get` implies it retrieves and returns something which this function doesn't. This function outputs the logs so the name should reflect that. We could rename to `outputPodLogs`.

tpantelis commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -62,41 +61,46 @@ func gatherPodLogsByContainer(podLabelSelector, container string, info Info) {
+	return nil
+}
+
+func getLogsFromStream(logStream io.ReadCloser) (string, error) {
```
```suggestion
func getLogFromStream(logStream io.ReadCloser) (string, error) {
```

tpantelis commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -62,41 +61,46 @@ func gatherPodLogsByContainer(podLabelSelector, container string, info Info) {
 
-func processLogStream(logrequest *rest.Request) (string, error) {
-	logStream, err := logrequest.Stream()
+func writeLogs(logStream io.ReadCloser, info Info, podName, fileExtension string) error {
```
```suggestion
func writePodLogToFile(logStream io.ReadCloser, info Info, podName, fileExtension string) error {
```

tpantelis commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -120,3 +124,39 @@ func findPods(clientSet kubernetes.Interface, byLabelSelector string) (*corev1.P
 	return pods, nil
 }
+
+func getPreviousLogs(pod *corev1.Pod, podLogOptions *corev1.PodLogOptions, info Info) error {
```
Similar to above, this doesn't retrieve/return something. Also this mutates the `podLogOptions` param so pass by value.

```suggestion
func outputPreviousPodLog(pod *corev1.Pod, podLogOptions corev1.PodLogOptions, info Info) error {
```

tpantelis commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -120,3 +124,39 @@ func findPods(clientSet kubernetes.Interface, byLabelSelector string) (*corev1.P
+	return nil
+}
+
+func getExistingLogs(pod *corev1.Pod, podLogOptions *corev1.PodLogOptions, info Info) error {
```
```suggestion
func outputCurrentPodLog(pod *corev1.Pod, podLogOptions corev1.PodLogOptions, info Info) error {
```

tpantelis commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -120,3 +124,39 @@ func findPods(clientSet kubernetes.Interface, byLabelSelector string) (*corev1.P
+	// if no previous pods found, logstream == nil, ignore it
+	if logStream != nil {
+		info.Status.QueueWarningMessage(fmt.Sprintf("found previous instances of pod %s", pod.Name))
+		err := writeLogs(logStream, info, pod.Name, ".log.old")
```
"_.old_" to me kind of implies a copy of another file.  I suggest "_.log.prev_".

tpantelis commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -120,3 +124,39 @@ func findPods(clientSet kubernetes.Interface, byLabelSelector string) (*corev1.P
+	logRequest := info.ClientSet.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, podLogOptions)
+	logStream, _ := logRequest.Stream()
+
+	// TODO: Check for error other than "no previous pods found"
```
yeah we shouldn't ignore the error.

tpantelis replied:
What about the TODO here? It's not a big deal - I guess we can deal with it later.

tpantelis commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -120,3 +124,39 @@ func findPods(clientSet kubernetes.Interface, byLabelSelector string) (*corev1.P
+
+	// if no previous pods found, logstream == nil, ignore it
+	if logStream != nil {
+		info.Status.QueueWarningMessage(fmt.Sprintf("found previous instances of pod %s", pod.Name))
```
```suggestion
		info.Status.QueueWarningMessage(fmt.Sprintf("Found logs for previous instances of pod %s", pod.Name))
```

mangelajo commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -62,41 +61,45 @@ func gatherPodLogsByContainer(podLabelSelector, container string, info Info) {
+func outputPodLogs(pod *corev1.Pod, podLogOptions corev1.PodLogOptions, info Info) error {
+	err := outputPreviousPodLog(pod, podLogOptions, info)
+	if err != nil {
+		return err
```
The problem here is that if this one fails the next one won’t even be attempted. I would save the error and return an aggregated error at the end of the function so we don’t skip the other log dumper.

Jaanki replied:
yes. Good catch @mangelajo

tpantelis commented on pkg/subctl/cmd/gather/logs.go:
```diff
@@ -62,41 +61,55 @@ func gatherPodLogsByContainer(podLabelSelector, container string, info Info) {
+	for _, err := range outputPodLogsErrors {
+		outputPodLogsError += err + "\n" // Prints each error in a new line
+	}
+	return errors.New(outputPodLogsError)
```
Rather than concatenating errors, I suggest to simplify it and just output the errors directly:

```
func outputPodLogs(pod *corev1.Pod, podLogOptions corev1.PodLogOptions, info Info) {
    err := outputPreviousPodLog(pod, podLogOptions, info)
    if err != nil {
        info.Status.QueueFailureMessage(fmt.Sprintf("Error outputting previous log for pod %q: %v", pod.Name, err))		
    }

    err := outputCurrentPodLog(pod, podLogOptions, info)
    if err != nil {
        info.Status.QueueFailureMessage(fmt.Sprintf("Error outputting current log for pod %q: %v", pod.Name, err))
    }
}
```
